cms/bundles/models.py

```diff
@@ -0,0 +1,201 @@
+
+    objects = BundleManager()
+
+    panels: ClassVar[list["Panel"]] = [
+        FieldPanel("name"),
+        FieldRowPanel(
+            [
+                FieldPanel("release_calendar_page", heading="Release Calendar page"),
+                FieldPanel("publication_date", heading="or Publication date"),
+            ],
+            heading=_("Scheduling"),
+            icon="calendar",
+        ),
+        FieldPanel("status"),
+        InlinePanel("bundled_pages", heading=_("Bundled pages"), icon="doc-empty", label=_("Page")),
+        # these are handled by the form
+        FieldPanel("approved_by", classname="hidden w-hidden"),
+        FieldPanel("approved_at", classname="hidden w-hidden"),
+    ]
+
+    search_fields: ClassVar[list[index.SearchField | index.AutocompleteField]] = [
+        index.SearchField("name"),
+        index.AutocompleteField("name"),
+    ]
+
+    def __str__(self) -> str:
+        return str(self.name)
+
+    @cached_property
+    def scheduled_publication_date(self) -> Optional["datetime.datetime"]:
+        """Returns the direct publication date or the linked release calendar page, if set."""
+        date: datetime.datetime | None = self.publication_date
+        if not date and self.release_calendar_page_id:
+            date = self.release_calendar_page.release_date  # type: ignore[union-attr]
+        return date
+
+    @property
+    def can_be_approved(self) -> bool:
+        """Determines
+        Note: strictly speaking, the bundle should in "in review" in order for it to be approved.
+        """
+        return self.status in [BundleStatus.PENDING, BundleStatus.IN_REVIEW]
+
+    def get_bundled_pages(self) -> QuerySet[Page]:
+        """Convenience method to return all bundled pages."""
+        pages: QuerySet[Page] = Page.objects.filter(pk__in=self.bundled_pages.values_list("page__pk", flat=True))
+        return pages
+
+    def save(self, **kwargs: Any) -> None:  # type: ignore[override]
+        """Adds additional behaviour on bundle saving.
+
+        For non-released bundles, we update the publication date for related pages if needed.
+        """
+        super().save(**kwargs)
+
+        if self.status == BundleStatus.RELEASED:
+            return
+
+        if self.scheduled_publication_date and self.scheduled_publication_date >= now():
+            # Schedule publishing for related pages.
+            # ignoring [attr-defined] because Wagtail is not fully typed and mypy is confused.
+            for bundled_page in self.get_bundled_pages().defer_streamfields().specific():  # type: ignore[attr-defined]
+                if bundled_page.go_live_at == self.scheduled_publication_date:
+                    continue
+
+                # note: this could use a custom log action for history
+                bundled_page.go_live_at = self.scheduled_publication_date
+                revision = bundled_page.save_revision()
```

Comment:
no, because save_revision iterates through all fields and accesses them. I do want to rework this to filter out bundled pages with go_live_at == scheduled_publication_date